R/zchunk_L244.building_det.R

```diff
@@ -0,0 +1,1044 @@
+      mutate(degree.days = round(value, energy.DIGITS_HDDCDD))
+
+    # L244.GenericServiceSatiation: Satiation levels assumed for non-thermal building services
+    # First, calculate the service output per unit floorspace in the USA region
+    L244.ServiceSatiation_USA <- L144.base_service_EJ_serv %>%
+      filter(GCAM_region_ID == gcam.USA_CODE) %>%
+      # Using left_join_keep_first_only b/c there are repeats and we only need to keep generic columns
+      left_join_keep_first_only(calibrated_techs_bld_det, by = c("sector", "service")) %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID") %>%
+      select(region, sector, service, gcam.consumer, nodeInput, building.node.input, building.service.input = supplysector, year, value)
+
+    # Floorspace should be matched in for a specified year, from the full floorspace table (i.e. not one that is subsetted to model base years)
+    L144.flsp_bm2_R_res_Yh <- L144.flsp_bm2_R_res_Yh %>%
+      mutate(gcam.consumer = A44.gcam_consumer$gcam.consumer[grepl("res", A44.gcam_consumer$gcam.consumer)]) %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID")
+
+    L144.flsp_bm2_R_comm_Yh <- L144.flsp_bm2_R_comm_Yh %>%
+      mutate(gcam.consumer = A44.gcam_consumer$gcam.consumer[grepl("com", A44.gcam_consumer$gcam.consumer)]) %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID")
+
+    L244.flsp_bm2_R <- bind_rows(L144.flsp_bm2_R_res_Yh, L144.flsp_bm2_R_comm_Yh) %>%
+      # Again, used to be energy.SATIATION_YEAR, changed to pass timeshift test
+      filter(year == max(BASE_YEARS)) %>%
+      select(-year)
+
+    L244.ServiceSatiation_USA <- L244.ServiceSatiation_USA %>%
+      left_join_error_no_match(L244.flsp_bm2_R %>%
+                                 rename(floorspace_bm2 = value), by = c("region", "gcam.consumer")) %>%
+      left_join_error_no_match(A44.demand_satiation_mult, by = c("building.service.input" = "supplysector")) %>%
+      group_by(region, sector, service) %>%
+      mutate(satiation.level = round(value[year == max(HISTORICAL_YEARS)] * multiplier / floorspace_bm2, energy.DIGITS_CALOUTPUT)) %>%
+      ungroup() %>%
+      select(-multiplier)
+
+    # Generic services: read these values to all regions because they're all the same
+    L244.GenericServiceSatiation <- L244.ServiceSatiation_USA %>%
+      filter(building.service.input %in% generic_services) %>%
+      select(-region, -GCAM_region_ID, -year, -value) %>%
+      distinct() %>%
+      write_to_all_regions(LEVEL2_DATA_NAMES[["GenericServiceSatiation"]], GCAM_region_names = GCAM_region_names)
+
+    # This is bad. Need to match in the floorspace into the base service table, divide to calculate the service demand
+    # per unit floorspace in the final calibration year. This (increased slightly) is then the minimum satiation level that needs to be read in.
+    # TODO: fix the bad code in the model. need a flexible building service function
```

Comment:
Is this your TODO, or Page's?

Reply:
This is and below are both Page's. Should we keep them?

Reply:
Let's move them into issues please, referencing file and line numbers.

Reply:
No need to repeat that text exactly in any code or issues; basically what's going on here is we have to jump through a bunch of hoops on the data side to make up for inflexible functional forms in the model. That said, developing new equations describing the buildings sector and re-coding the thing is probably not going to be a priority in the next few years. Dunno if it's worth an issue on github, Jira, or neither.

Reply:
You call I guess @pkyle - but if in doubt, please do capture in an issue. @russellhz - let's delete the TODO text from the code please.